smallrye-reactive-messaging-mqtt-server/src/main/java/io/smallrye/reactive/messaging/mqtt/server/MqttServerSource.java

```diff
@@ -0,0 +1,159 @@
+import static io.vertx.core.net.NetworkOptions.DEFAULT_RECEIVE_BUFFER_SIZE;
+import static io.vertx.core.net.TCPSSLOptions.DEFAULT_SSL;
+import static io.vertx.mqtt.MqttServerOptions.DEFAULT_MAX_MESSAGE_SIZE;
+import static io.vertx.mqtt.MqttServerOptions.DEFAULT_PORT;
+import static io.vertx.mqtt.MqttServerOptions.DEFAULT_TIMEOUT_ON_CONNECT;
+import static io.vertx.mqtt.MqttServerOptions.DEFAULT_TLS_PORT;
+
+import java.util.concurrent.CompletableFuture;
+
+import org.eclipse.microprofile.config.Config;
+import org.eclipse.microprofile.reactive.streams.operators.PublisherBuilder;
+import org.eclipse.microprofile.reactive.streams.operators.ReactiveStreams;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import io.reactivex.processors.BehaviorProcessor;
+import io.vertx.mqtt.MqttServerOptions;
+import io.vertx.reactivex.core.Vertx;
+import io.vertx.reactivex.mqtt.MqttServer;
+
+class MqttServerSource {
+
+    private final Logger logger = LoggerFactory.getLogger(MqttServerSource.class);
+    private final boolean broadcast;
+    private final PublisherBuilder<MqttMessage> source;
+    private final MqttServer mqttServer;
+
+    private static MqttServerOptions mqttServerOptions(Config config) {
+        final MqttServerOptions options = new MqttServerOptions();
+        options.setAutoClientId(config.getOptionalValue("auto-client-id", Boolean.class).orElse(true));
+        options.setSsl(
+                config.getOptionalValue("ssl", Boolean.class).orElse(DEFAULT_SSL));
+        // TODO set KeyCertOptions if SSL, c.f. https://vertx.io/docs/vertx-mqtt/java/#_handling_client_connection_disconnection_with_ssl_tls_support
+        options.setMaxMessageSize(config.getOptionalValue("max-message-size", Integer.class)
+                .orElse(DEFAULT_MAX_MESSAGE_SIZE));
+        options.setTimeoutOnConnect(config.getOptionalValue("timeout-on-connect", Integer.class)
+                .orElse(DEFAULT_TIMEOUT_ON_CONNECT));
+        options.setReceiveBufferSize(config.getOptionalValue("receive-buffer-size", Integer.class)
+                .orElse(DEFAULT_RECEIVE_BUFFER_SIZE));
+        final int defaultPort = options.isSsl() ? DEFAULT_TLS_PORT : DEFAULT_PORT;
+        options.setPort(config.getOptionalValue("port", Integer.class).orElse(defaultPort));
+        options.setHost(
+                config.getOptionalValue("host", String.class).orElse(DEFAULT_HOST));
+        return options;
+    }
+
+    MqttServerSource(Vertx vertx, Config config) {
+        this.broadcast = config.getOptionalValue("broadcast", Boolean.class).orElse(false);
+        final MqttServerOptions options = mqttServerOptions(config);
+        this.mqttServer = MqttServer.create(vertx, options);
+        final BehaviorProcessor<MqttMessage> processor = BehaviorProcessor.create();
+
+        mqttServer.exceptionHandler(error -> {
+            logger.error("Exception thrown", error);
+            processor.onError(error);
```

Comment:
Do we want to retry? Might be something we can add later.

Reply:
In my case I was going to count on Kubernetes to restart it. But it makes sense, I'll look at it to see if I can quickly implement a retry mechanism.

Reply:
No hurry, what you did it what it generally intended.